### PR TITLE
Drop transactional table should take appropriate lock

### PR DESCRIPTION
**Pull request by harmandeeps**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveTransaction.java [hunk 1]
```diff
@@ -56,12 +57,12 @@ public long getTransactionId()
...
     {
         // Different calls for same table might need to lock different partitions so acquire locks every time
-        metastore.acquireSharedReadLock(
+        metastore.acquireLock(
                 identity,
                 queryId,
                 transactionId,
-                tableHandle.getPartitions().isEmpty() ? ImmutableList.of(tableHandle.getSchemaTableName()) : ImmutableList.of(),
-                tableHandle.getPartitions().orElse(ImmutableList.of()));
+                !tableHandle.getPartitions().isPresent() ? ImmutableList.of(tableHandle.getSchemaTableName()) : ImmutableList.of(),
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveTransaction.java:64]: revert

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 2]
```diff
@@ -1220,15 +1222,28 @@ private void commitShared()
...
 
-        private void prepareDropTable(HiveIdentity identity, SchemaTableName schemaTableName)
+        private void prepareDropTable(HdfsContext context, HiveIdentity identity, SchemaTableName schemaTableName)
         {
-            metastoreDeleteOperations.add(new IrreversibleMetastoreOperation(
-                    format("drop table %s", schemaTableName),
-                    () -> delegate.dropTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName(), true)));
+            Table table = delegate.getTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName())
+                    .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            if (AcidUtils.isTransactionalTable(table.getParameters())) {
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java:1234]: It feel that you can just wrap the operation into locking logic and use `metastoreDeleteOperations` list. Here.

Sth like this:
```java
    public Runnable runWithLock(HdfsContext context, HiveIdentity identity, SchemaTableName schemaTableName, Runnable action)
    {
        return () -> {
            long heartbeatInterval = configuredTransactionHeartbeatInterval
                    .map(Duration::toMillis)
                    .orElseGet(SemiTransactionalHiveMetastore.this::getServerExpectedHeartbeatIntervalMillis);
            long transactionId = delegate.openTransaction(identity);

            String queryId = context.getQueryId().orElseThrow(() -> new IllegalArgumentException("query ID not present"));
            log.debug(format("Opened transaction: %s for queryID: %s for table: %s", transactionId, queryId, schemaTableName.getTableName()));

            ScheduledFuture<?> heartbeatTask = heartbeatExecutor.scheduleAtFixedRate(
                    () -> delegate.sendTransactionHeartbeat(identity, transactionId),
                    0,
                    heartbeatInterval,
                    MILLISECONDS);

            delegate.acquireLock(
                    identity,
                    queryId,
                    transactionId,
                    ImmutableList.of(schemaTableName),
                    ImmutableList.of(),
                    DataOperationType.NO_TXN);

            try {
                action.run();
            }
            finally {
                heartbeatTask.cancel(true);
                delegate.commitTransaction(identity, transactionId);
            }
        };
    }
```

and then
```java
        private void prepareDropTable(HdfsContext context, HiveIdentity identity, SchemaTableName schemaTableName)
        {
            Table table = delegate.getTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName())
                    .orElseThrow(() -> new TableNotFoundException(schemaTableName));

            Runnable operation = () -> delegate.dropTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName(), true);
            if (AcidUtils.isTransactionalTable(table.getParameters())) {
                operation = runWithLock(context, identity, schemaTableName, operation);
            }

            metastoreDeleteOperations.add(new IrreversibleMetastoreOperation(
                    format("drop table %s", schemaTableName),
                    operation));
        }
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 3]
```diff
@@ -2587,6 +2609,68 @@ public void run()
...
+                    heartbeatInterval,
+                    MILLISECONDS);
+
+            delegate.acquireLock(
+                    identity,
+                    queryId,
+                    transactionId,
+                    ImmutableList.of(schemaTableName),
+                    ImmutableList.of(),
+                    DataOperationType.NO_TXN);
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java:2662]: why `NO_TXN`?

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java [hunk 4]
```diff
@@ -400,6 +410,82 @@ public void testFilesForAbortedTransactionsIgnored()
...
+
+            // Acquire exclusive lock manually on table
+            LockResponse lockResponse = transactionalTestHelper.acquireDropTableLock(client, "default", tableName, transactionId);
+            checkState(lockResponse.getState() == LockState.ACQUIRED, "Drop table didn't acquire exclusive lock");
+
+            String dropTable = "DROP TABLE " + tableName;
+            // Run DROP TABLE command
+            executor.execute(() -> defaultQueryExecutor().executeQuery(dropTable));
+
+            Thread.sleep(5000);
```
Comment (losipiuk) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:455]: This looks scary.

**Review by losipiuk — COMMENTED** [review]
skimmed

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 2 above)
Comment (harmandeeps) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java:1234]: The notion of creating `TransactionalMetastoreOperation` is to register actions which we want to run on a transactional table by acquiring a lock and then run the action. Right now, change is only for `drop table`, but we may be needing this for other DDL commands like `alter table` which need an exclusive lock.

The way `run` method is structured, keeping this into an account. We can register any transactional table action with it and `run` will take care of opening a transaction, creating heartbeat, acquiring lock, commit/rollingback transaction.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveTransaction.java (hunk 1 above)
Comment (harmandeeps) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveTransaction.java:64]: this change it needed to send correct `DataOperationType`.

**Comment by losipiuk** [comment]
Could you please rabase on master? :)

**Comment by mosabua** [comment]
:wave: @harmandeeps - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Pull request closed without merging**